### PR TITLE
Incorrect refund calculation

### PR DESCRIPTION
**Pull request by lukemorton**
I've added a failing test to request feedback on this potential incorrect refund calculation.

In the test for Spree::Calculator::Refunds::DefaultRefundAmount I have introduced a tax rate and then added an example that tests that had the full order been refunded, the correct amount would have been refunded. The test fails because even if we refunded both line items, the combined total with tax applied is less than the order total.

This _is_ a bug right?

Related to: #5915

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://madebymade-bugfix-5915-incorrect-refund-calculation.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://madebymade-bugfix-5915-incorrect-refund-calculation.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/spec/models/spree/calculator/refunds/default_refund_amount_spec.rb [hunk 1]
```diff
@@ -3,30 +3,43 @@
...
       before do
         order.adjustments << create(:adjustment, order: order, amount: adjustment_amount, eligible: true, label: 'Adjustment', source_type: 'Spree::Order')
         order.adjustments.first.update_attributes(amount: adjustment_amount)
+        order.updater.update_totals
       end
 
-      it { is_expected.to eq (pre_tax_amount - adjustment_amount.abs) / line_item_quantity }
+      it { is_expected.to eq (line_item.pre_tax_amount - adjustment_amount.abs) / line_item_quantity }
+
+      context 'refund amount * line item quantity with tax applied' do
```
Comment (athal7) [inline core/spec/models/spree/calculator/refunds/default_refund_amount_spec.rb]: can we have two different specs, one for inclusive tax and one for additional tax

Comment (DrPheltRight) [inline core/spec/models/spree/calculator/refunds/default_refund_amount_spec.rb]: Sure no problem!

Comment (DrPheltRight) [inline core/spec/models/spree/calculator/refunds/default_refund_amount_spec.rb]: Have now split out into two different contexts using shared_context.

**Comment by mvidaurre** [comment]
I having the same problem than you @DrPheltRight because in Mexico the VAT (IVA in Spanish) by law should be included in the price. Do you need any help for fixing this?

**Comment by JDutil** [comment]
There is still a broken spec:
rspec ./spec/models/spree/calculator/refunds/default_refund_amount_spec.rb:45

**Comment by JDutil** [comment]
Sorry didn't realize it was just adding the failing spec thought there was a fix as well.

**Comment by DrPheltRight** [comment]
I've identified the problem as the issue of adjustments being affected by inclusive tax. We require pretax amounts for adjustments that are based on taxed amounts (line item percentage, etc.) in order to calculate a pretax refund amount. Until we are able to identify which adjustments are affected by inclusive tax on products we will not be able to solve this bug.

**Comment by DrPheltRight** [comment]
I'm working on a solution that will filter adjustments by those where `Calculator#tax_affected? == true`. Then calculate inclusive tax and subtract that from the calculated refund amount.

**Comment by DrPheltRight** [comment]
Okay so here's my initial solution for fixing Spree::Calculator::Returns::DefaultRefund amount.

I have added #tax_affected? to calculators so in the default refund calculator we can work out the pre tax amount for adjustments that are affected by tax. This should the create the correct pretax amounts on Spree::ReturnItem's.

This change may well break the reimbursement implementation which I will start taking a look at next week.

Are we in consensus this is the right way to go?

Thanks all!

**Comment by athal7** [comment]
i'm not sure i understand why you have chosen those calculators to be 'tax_affected'. isn't it dependent on the store whether those calculated promotions are pre-tax or post-tax? not inherent in the promotion itself?

**Comment by DrPheltRight** [comment]
Thanks @athal7 for getting back to me on this. I chose to mark calculators as #tax_affected? since it's up to the calculator authors to say whether these calculators are affected by tax.

The default refund calculator then applies inclusive tax reductions on the adjustment total. This means that if the calculator is affected by tax but the order has 0 inclusive tax then the adjustment is not changed.

**Comment by athal7** [comment]
i think i'd rather defer to the calculator's choice rather than it being explicit on the promotion action, otherwise you would need different promotion actions depending on where you plan on selling the item, and the tax_inclusive setting is already administered on the tax_rate

**Comment by DrPheltRight** [comment]
The choice is deferred to the calculator:

https://github.com/madebymade/spree/blob/79cca7e92368cccc575823834d6450707d586c35/core/app/models/concerns/spree/calculated_adjustments.rb#L10

**Comment by athal7** [comment]
i meant the tax calculator, not the promotion calculator

**Comment by DrPheltRight** [comment]
I do not see the need to have this as configurable from the admin. What in particular would be broken by this suggested implementation?

**Comment by athal7** [comment]
Sorry that's not quite what I meant. It seems that the taxability of the adjustment ought to be knowledge of the tax rate calculator, not the promotion calculator, since it is ultimately the tax rate calculator that is implementing whether it is pre- or post- tax. With this approach, if you are selling some items pre-tax and some items post-tax, you need different promotion calculators for those different items, which seems suboptimal. Maybe I am misunderstanding something.

**Comment by JDutil** [comment]
I also don't think having tax_affected? method for each calculator makes sense when tax inclusion is set on the tax rate.  When a line item or order is passed to the calculators you should be able to check if there is any included_tax_total set on the object, and if so then it would be tax affected.

**Comment by simmsy** [comment]
Guys. I think the bug maybe here:
https://github.com/spree/spree/blob/master/core/app/models/spree/return_item.rb#L123:L125

It should be:
`pre_tax_amount + included_tax_total + additional_tax_total`

From what I've discovered the TaxCalculator above was working correctly but quite simply the included_tax_total was not added when calculating the total to reimburse

**Comment by JDutil** [comment]
Nice catch @simmsy

**Comment by JDutil** [comment]
I've tried implementing both @simmsy suggestion and the spec changes here, and they still fail... Going to investigate some more but not so sure at this point whether it's just the specs are bad or if the change solves the problem as I expected it too.

**Pull request closed without merging**